apps/sync/js/sync-engine/syncengine.js

```diff
@@ -0,0 +1,358 @@
+ *   syncing.
+ * - These collections are set up with the correct IdSchema (typically
+ *   allowing only 9-character URL-safe Base64 strings, although for some
+ *   collections other strings are allowed, e.g. bookmarks also has id's like
+ *   'menu' and 'toolbar').
+ * - Adding a RemoteTransformer to each of these collections, so that outgoing
+ *   and incoming data is passed through FxSyncWebCrypto for its en-/decryption.
+ * - For each collection to be synced, the following steps occur:
+ *   - Download the latest remote data into the local decrypted copy, managed by
+ *     kinto.js.
+ *   - For each incoming conflict reported by kinto.js (if any), call the
+ *     `handleConflict` method on the DataAdapter registered for this
+ *     collection.
+ *   - When kinto.js has obtained a full decrypted copy of the remote data, call
+ *     the `update` method on the DataAdapter, passing it an interface to the
+ *     Kinto collection in question.
+ *   - When this method returns its promise, sync the Kinto collection a second
+ *     time, to make sure any changes made there by the DataAdapter are synced
+ *     up to the remote Firefox Sync server (again, through the remote Syncto
+ *     proxy server). The RemoteTransformer on the Kinto collection will take
+ *     care of encrypting the records before they get uploaded.
+ *   - Call the `handleConflict` method on the DataAdapter again for any
+ *     conflicts kinto.js may report during this second sync process.
+ *
+ */
+
+'use strict';
+
+/* global
+  crypto,
+  FxSyncWebCrypto,
+  Kinto
+*/
+
+/* exported
+  SyncEngine
+*/
+
+var SyncEngine = (function() {
+  var FxSyncIdSchema = Kinto.createIdSchema({
+    constructor: function(collectionName) {
+      this.collectionName = collectionName;
+    },
+    generate: function() {
+      var bytes = new Uint8Array(9);
+      crypto.getRandomValues(bytes);
+      var binStr = '';
+      for (var i = 0; i < 9; i++) {
+          binStr += String.fromCharCode(bytes[i]);
+      }
+      // See https://docs.services.mozilla.com/storage/apis-1.5.html
+      return window.btoa(binStr).replace('+', '-').replace('/', '_');
+    },
+    validate: function(id) {
+      // FxSync id's should be 12 ASCII characters, representing 9 bytes of data
+      // in modified Base64 for URL variants exist, where the '+' and '/'
+      // characters of standard Base64 are respectively replaced by '-' and '_'
+      // See https://docs.services.mozilla.com/storage/apis-1.5.html
+      return /^[A-Za-z0-9-_]{12}$/.test(id);
+    }
+  });
+
+  var ControlCollectionIdSchema = Kinto.createIdSchema({
+    constructor: function(collectionName, keyName) {
+      this.collectionName = collectionName;
+      this.keyName = keyName;
+    },
+    generate: function() {
+      return this.keyName;
+    },
+    validate: function(id) {
+      return (id === this.keyName);
+    }
+  });
+
+  var WebCryptoTransformer = Kinto.createRemoteTransformer({
```

Comment:
I wonder if we need to create this even if there are no changes locally or in the server.